axelor-account/src/main/java/com/axelor/apps/account/service/invoice/InvoiceLineServiceImpl.java

```diff
@@ -452,4 +458,82 @@ public Unit getUnit(Product product, boolean isPurchase) {
 
     return productInformation;
   }
+
+  @Override
+  public boolean hasEndOfPackTypeLine(List<InvoiceLine> invoiceLineList) {
+    return ObjectUtils.isEmpty(invoiceLineList)
+        ? Boolean.FALSE
+        : invoiceLineList.stream()
+            .anyMatch(
+                invoiceLine ->
+                    invoiceLine.getTypeSelect() == InvoiceLineRepository.TYPE_END_OF_PACK);
+  }
+
+  @Override
+  public boolean isStartOfPackTypeLineQtyChanged(List<InvoiceLine> invoiceLineList) {
+
+    if (ObjectUtils.isEmpty(invoiceLineList)) {
+      return false;
+    }
+    for (InvoiceLine invoiceLine : invoiceLineList) {
+      if (invoiceLine.getTypeSelect() == InvoiceLineRepository.TYPE_START_OF_PACK
+          && invoiceLine.getId() != null) {
+        InvoiceLine oldInvoiceLine = invoiceLineRepo.find(invoiceLine.getId());
+        if (oldInvoiceLine.getTypeSelect() == InvoiceLineRepository.TYPE_START_OF_PACK
+            && invoiceLine.getQty().compareTo(oldInvoiceLine.getQty()) != 0) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public InvoiceLine updateProductQty(
+      InvoiceLine invoiceLine, Invoice invoice, BigDecimal oldQty, BigDecimal newQty, int scale) {
```

Comment:
I think it would be better to remove scale from method argument and declare variable for it inside method.